src/userdb/userdbctl.c

```diff
@@ -1164,6 +1169,220 @@ static int ssh_authorized_keys(int argc, char *argv[], void *userdata) {
+
+                r = write_string_file_at(userdb_dir_fd, fn, formatted, WRITE_STRING_FILE_CREATE|WRITE_STRING_FILE_ATOMIC|WRITE_STRING_FILE_MODE_0600);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to write JSON record to /run/userdb/%s: %m", fn);
+
+                free(link);
+
+                if (ur) {
+                        if (asprintf(&link, UID_FMT ".user-privileged", ur->uid) < 0)
+                                return log_oom();
+                } else {
+                        if (asprintf(&link, GID_FMT ".group-privileged", gr->gid) < 0)
+                                return log_oom();
+                }
+
+                if (symlinkat(fn, userdb_dir_fd, link) < 0)
+                        return log_error_errno(errno, "Failed to symlink %s to %s", link, fn);
+
+                log_info("Installed /run/userdb/%s from credential.", fn);
+        }
+
+        if (ur) {
+                STRV_FOREACH(g, ur->member_of) {
+                        _cleanup_free_ char *membership = strjoin(ur->user_name, ":", *g);
+                        if (!membership)
+                                return log_oom();
+
+                        _cleanup_close_ int fd = openat(userdb_dir_fd, membership, O_WRONLY|O_CREAT|O_CLOEXEC, 0644);
+                        if (fd < 0)
+                                return log_error_errno(errno, "Failed to create %s: %m", membership);
+
+                        log_info("Installed /run/userdb/%s from credential.", membership);
+                }
+        } else {
+                STRV_FOREACH(u, gr->members) {
+                        _cleanup_free_ char *membership = strjoin(*u, ":", gr->group_name);
+                        if (!membership)
+                                return log_oom();
+
+                        _cleanup_close_ int fd = openat(userdb_dir_fd, membership, O_WRONLY|O_CREAT|O_CLOEXEC, 0644);
+                        if (fd < 0)
+                                return log_error_errno(errno, "Failed to create %s: %m", membership);
+
+                        log_info("Installed /run/userdb/%s from credential.", membership);
+                }
+        }
+
+        if (ur && user_record_disposition(ur) == USER_REGULAR) {
+                const char *hd = user_record_home_directory(ur);
+
+                r = RET_NERRNO(access(hd, F_OK));
+                if (r < 0) {
+                        if (r != -ENOENT)
+                                return log_error_errno(r, "Failed to check if %s exists: %m", hd);
+
+                        r = mkdir_parents(hd, 0755);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to create parent directories of %s: %m", hd);
+
+                        if (mkdir(hd, 0700) < 0 && errno != EEXIST)
+                                return log_error_errno(r, "Failed to create %s: %m", hd);
```

Comment:
probably makes sense to use WITH_UMASK(000) here...

Reply:
r -> errno